### PR TITLE
feat(evals): group trajectories by run and cross-link Braintrust experiments

### PR DESCRIPTION
**Pull request by miguelg719**
## What

Stacked on #2138.

- Persist trajectories under `<root>/<experiment>__<model>/<task.id>/<runId>/` instead of scattered per-task timestamp dirs, so a run's trajectories live in one folder and concurrent multi-model runs of the same suite never interleave on disk.
- Write a `metadata.json` into every trajectory dir (experiment, model, provider, environment, task, runId, status) so a trajectory's run never has to be reverse-engineered after the fact.
- Write an `experiment.json` at the group root cross-linking the local trajectories to the resolved Braintrust experiment (hashed name, id, project, URLs) once `Eval()` resolves.
- New `framework/trajectoryGroup.ts` owns the slugging/layout/metadata helpers; `TrajectoryRecorder`, the external-harness persister, and both eval entrypoints (`index.eval.ts`, `framework/runner.ts`) consume it.

## Why

Mapping on-disk trajectories back to their Braintrust experiment previously required guessing by timestamp, and concurrent runs of different models wrote into the same root with no on-disk marker — making post-hoc analysis error-prone.

## Notes

- Local persistence only; Braintrust experiment naming/metadata is unchanged.
- Tooling that globs `.trajectories/*` keeps working (one extra path level); the trajectoryRecorder unit test asserts the new layout + `metadata.json`.

---
## Summary by cubic
Groups eval trajectories by experiment, unambiguous model, and a run-unique token, and cross-links each run to its Braintrust experiment. Adds atomic dir reservation, single-flight caching, and env hygiene so runs never overwrite, mislabel, or leak; the run token uses 64‑bit entropy and result persistence is order-free (also trims internal comments; no logic changes).

- **New Features**
  - Store trajectories at `<root>/<experiment>[__<model>]__<runToken>/<task.id>/<runId>/`; write `metadata.json` per trajectory and an `experiment.json` at the group root linking to the resolved Braintrust experiment; implemented in `framework/trajectoryGroup.ts` and used by the recorder, adapter persister, runner, and `index.eval.ts`.
  - Generate one run token per run (64‑bit entropy); derive the group/metadata model from the actual testcase matrix when unambiguous (`EVAL_TRAJECTORY_MODEL`), not the requested override; stamp `EVAL_EXPERIMENT_NAME`/`EVAL_TRAJECTORY_GROUP`; `writeExperimentLink()` takes an explicit group, respects `EVAL_TRAJECTORY_ROOT`, and only writes if persistence is on and the group dir exists.
  - Bench hardening: add `taskDates.ts` for safe rolling dates (WebMD); update Healthline to stop before final subscribe; clarify Google Maps rubric; KFC requires pickup scheduling; SFPL uses the live application URL; default `maxSteps`: `agent/sf_library_card` 1→10, `agent/sf_library_card_multiple` 20→25. TUI `verify` help and docs reflect the new layout.

- **Bug Fixes**
  - Prevent clobbering by atomically reserving trajectory dirs (`reserveTrajectoryDir` adds `-2/-3` on collision) with single-flight caching in the recorder; `finish()` is idempotent; result persistence is order-free (now covered by tests); `metadata.json` records `runDir` and `attempt`.
  - Hardened slugging: reject pure-dot values (e.g., `"."`, `".."`) to avoid path escapes; fall back to the `"default"` group.
  - REPL env hygiene: `withEnvOverrides()` snapshots/restores run-stamped env (`EVAL_TRAJECTORY_GROUP`, `EVAL_EXPERIMENT_NAME`, `EVAL_TRAJECTORY_MODEL`) so runs don’t leak between commands; recorder and adapter both honor `EVAL_TRAJECTORY_ROOT`.
  - Trimmed internal comments and moved a misplaced docstring; no logic changes.

<sup>Written for commit 2d17642f63e4570a4f9fbd167c455502de3624bb. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/browserbase/stagehand/pull/2177?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: packages/evals/index.eval.ts [hunk 1]
```diff
@@ -545,6 +559,21 @@ const generateFilteredTestcases = (): Testcase[] => {
       };
     });
 
+    // Cross-link local trajectories to the resolved Braintrust experiment.
+    // The hashed name (e.g. `agent/onlineMind2Web-92918006`) is only known now,
+    // after Eval() resolves — so write it once at the group-dir root.
+    const summary = evalResult.summary;
+    if (summary) {
+      await writeExperimentLink(resolveTrajectoryRoot(), {
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/evals/index.eval.ts]: Valid — `TrajectoryRecorder` defaulted to `<cwd>/.trajectories` and ignored the override. Fixed: the recorder's default root now goes through `resolveTrajectoryRoot()`, the same resolution the experiment-link write uses.

File: packages/evals/framework/trajectoryRecorder.ts [hunk 2]
```diff
@@ -212,13 +248,38 @@ export class TrajectoryRecorder {
...
+   * Reserve this recorder's on-disk directory exactly once. Reservation
+   * happens at first persistence (not construction) so collision resolution
+   * sees dirs concurrent recorders have actually created; the cached result
+   * keeps finish() idempotent and makes finish()/persistResult() order-free.
+   */
+  private async ensureReserved(): Promise<{
+    directory: string;
+    attempt: number;
+  }> {
+    if (!this.reserved) {
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/evals/framework/trajectoryRecorder.ts]: Valid, and it was **not** actually fixed by \`016c74e\` — that is the commit that introduced the code. \`ensureReserved()\` was still check-then-await.

Confirmed by mutation test: reverting to the old implementation makes the new test fail with

```
AssertionError: expected [ 'run-1', 'run-1-2' ] to deeply equal [ 'run-1' ]
```

so a single recorder really did split its artifacts across two run dirs. It was not firing in production only because `verifierAdapter` awaits `finish()` before `persistResult()` — nothing enforced that.

Fixed in `f3e4824` with the single-flight promise you suggested: the in-flight **promise** is cached (assigned synchronously, before any await), so overlapping callers share one reservation. A rejection clears the cache so a later attempt can still retry. Regression test added.

File: packages/evals/tasks/bench/agent/webmd_ovulation_calculator.ts [hunk 3]
```diff
@@ -15,8 +15,22 @@ export default defineBenchTask(
       const page = v3.context.pages()[0];
       await page.goto(initUrl);
 
-      const instruction =
-        "Search for the ovulation calculator and enter Mar 1 as the first date of the period and calculate the date of ovulation and pregnancy test day. use https://www.webmd.com/ to achieve the task. Don't go to any other site. The task is achievable with just navigation from this site.";
+      // WebMD's ovulation calculator only accepts a last-period start date
+      // within roughly the past 90 days, so a hardcoded date (previously
+      // "Mar 1") becomes task-invalid once >90 days elapse. Use the first day
+      // of the previous month, which is always within range (≤ ~62 days ago).
+      const periodDate = new Date();
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/evals/tasks/bench/agent/webmd_ovulation_calculator.ts]: Valid — fixed in `f3e4824`. Extracted the date arithmetic into `framework/taskDates.ts` and added tests covering both traps you implied: month-end overflow (`setMonth()` on Mar 31 asks for "Feb 31" and overflows forward into March) and the January rollback to December of the *prior* year. Plus a sweep asserting the generated date is always in the past and inside WebMD's ~90-day window — which is the invariant whose violation made the old hardcoded "Mar 1" task-invalid in the first place.

File: packages/evals/framework/trajectoryRecorder.ts [hunk 4]
```diff
@@ -212,13 +248,38 @@ export class TrajectoryRecorder {
...
   async persistResult(
     result: EvaluationResult,
     filename = "result.json",
   ): Promise<void> {
     if (!this.persistEnabled) return;
 
-    const scoresDir = path.join(this.outputDir, "scores");
+    // Route through the shared reservation so scores land in the same dir as
+    // the trajectory regardless of finish()/persistResult() call order.
+    const { directory } = await this.ensureReserved();
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/evals/framework/trajectoryRecorder.ts:291]: Does not reproduce — leaving as-is. `persistResult()` writes `<dir>/scores/result.json`, while `finish()` → `writeTrajectoryDir()` writes `task_data.json`, `trajectory.json` and `screenshots/` and never touches `scores/`. The `result` keys `writeTrajectoryDir` does write are `step.toolOutput.result` (per-step tool output), which is unrelated to the `EvaluationResult` that `persistResult()` persists — so a `persistResult()`-before-`finish()` ordering would not drop it.

Independently, the order is never reversed at the only call site (`verifierAdapter.ts`: `await recorder.finish(...)` then `await recorder.persistResult(...)`). The related *real* issue was the shared reservation, which is fixed in `f3e4824`.

**Comment by miguelg719** [comment]
@cubic-dev-ai

File: packages/evals/index.eval.ts [hunk 5]
```diff
@@ -341,6 +347,20 @@ const generateFilteredTestcases = (): Testcase[] => {
...
 
+  // Stamp the run-scoped trajectory group so every task in this run lands under
+  // one folder (`<root>/<experiment>__<model>__<runToken>/...`) instead of
+  // scattered per-task timestamps. The run token is generated ONCE here and
+  // reused at completion time, so re-running the same suite can't clobber the
+  // previous run's experiment.json. Local persistence only — does not affect
+  // Braintrust.
+  const trajectoryGroup = buildTrajectoryGroupSlug({
+    experimentName,
+    model: process.env.EVAL_MODEL_OVERRIDE,
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/evals/index.eval.ts]: Valid, and worse than described — fixed in `0b8c03e7`.

Confirmed the root cause: `cli-legacy.ts:615-617` sets **only** `EVAL_MODEL_OVERRIDE`, and `getModelList()` (`taskConfig.ts:325`) never reads it — it returns `getDefaultAgentModels()` / provider-filtered / `DEFAULT_EVAL_MODELS` (which reads `EVAL_MODELS`, a different var). So the override and the actual matrix are fully independent: `--model X` labelled every trajectory `model: X` while the testcases really ran gemini-2.5-flash, gpt-4.1-mini and claude-haiku-4-5.

Fix: new `resolveUnambiguousModel()` derives the model from the testcases that will actually run, and omits it unless it is unambiguous for the whole run. Metadata now reads a new `EVAL_TRAJECTORY_MODEL` (the *resolved* model) rather than the merely-*requested* override, and is cleared when ambiguous — absent beats wrong. Applied to both entrypoints.

File: packages/evals/framework/trajectoryGroup.ts [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+ * A compact, sortable, filesystem-safe token identifying a single run, e.g.
+ * "20260715-110342" (local time, YYYYMMDD-HHMMSS). Generate this EXACTLY ONCE
+ * per run in the entrypoint and reuse it — calling it twice within one run would
+ * split that run's trajectories across two group dirs.
+ */
+export function generateRunToken(now: Date = new Date()): string {
+  const pad = (n: number) => String(n).padStart(2, "0");
+  const date = `${now.getFullYear()}${pad(now.getMonth() + 1)}${pad(now.getDate())}`;
+  const time = `${pad(now.getHours())}${pad(now.getMinutes())}${pad(now.getSeconds())}`;
+  return `${date}-${time}`;
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/evals/framework/trajectoryGroup.ts]: Valid — fixed in `0b8c03e7`. Second-granularity was exactly the hole: two same-suite runs starting in the same second shared a group and clobbered each other's `experiment.json`, reintroducing the bug the token was added to prevent. `generateRunToken` now appends random entropy (`20260716-110342-9f3a1c`) while keeping the timestamp prefix lexicographically sortable. Regression test asserts two calls with the *same* injected `now` differ — it fails if the entropy is removed.

File: packages/evals/framework/trajectoryGroup.ts [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+ * No-ops when trajectory persistence is off (CI, core-tier runs), so those runs
+ * don't leave behind an otherwise-empty `.trajectories/<group>/` tree. The gate
+ * lives here — the single choke point — so a future caller can't forget it.
+ */
+export async function writeExperimentLink(
+  root: string,
+  group: string,
+  link: Record<string, unknown>,
+): Promise<void> {
+  if (!shouldPersistTrajectory(undefined)) return;
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/evals/framework/trajectoryGroup.ts]: Valid — fixed in `0b8c03e7`, and generalized. The runner now passes an explicit `{ persist: !hasCoreOnly && shouldPersistTrajectory(undefined) }` as you suggested.

But gating on the tier signal alone would have left the same symptom on the legacy entrypoint, which also runs non-agent categories (act/extract/combination) that never construct a `TrajectoryRecorder`. So `writeExperimentLink` additionally requires the group dir to already exist: it is created by the first trajectory reservation, so its absence means nothing was recorded. That is a universal backstop needing no tier signal, and it no longer `mkdir`s the group itself.

File: packages/evals/framework/harnesses/persistTrajectory.ts [hunk 8]
```diff
@@ -16,7 +22,9 @@ export interface PersistAdapterTrajectoryOptions {
   /** EvaluationResult from V3Evaluator.verify(). Written to scores/result.json. */
   evaluationResult?: EvaluationResult;
   /**
-   * Output directory root. Final layout lives at `<outputRoot>/<runId>/<task.id>/`.
+   * Output directory root. Final layout lives at
+   * `<outputRoot>/<group>/<task.id>/<runId>/` (group = EVAL_TRAJECTORY_GROUP,
+   * i.e. `<experiment>__<model>__<runToken>`, or "default").
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/evals/framework/harnesses/persistTrajectory.ts]: Valid — fixed in `0b8c03e7`. Reworded to describe `<experiment>[__<model>]__<runToken>` as the normal entrypoint-generated form (model present only when unambiguous) rather than an equivalence, and noted that integrations may set any group value.

File: packages/evals/tests/framework/trajectoryGroup.test.ts [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    expect(
+      buildTrajectoryGroupSlug({
+        experimentName: "agent",
+        model: "openai/gpt-4.1-mini",
+        runToken: "20260715-110342",
+      }),
+    ).toBe("agent__openai_gpt-4.1-mini__20260715-110342");
+  });
+
+  it("drops empty parts (no model, no run token)", () => {
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/evals/tests/framework/trajectoryGroup.test.ts:115]: Valid — fixed in `0b8c03e7`, and it turned up a worse gap next to it. Added the `model`-without-`runToken` case.

While covering it I found the adjacent core-only test was **vacuous**: `resolveUnambiguousModel(["none","NONE"])` returns `undefined` from the `resolved.size === 1` check whether or not the `"none"` sentinel filter exists, so it proved nothing about the sentinel. Verified by executing both implementations. `["none"]` — the real core-only shape, since `runner.ts:125` stamps exactly `modelName: "none"` on every core testcase — is the only discriminating input, and it was missing. Both are now asserted.

File: packages/evals/index.eval.ts [hunk 10]
```diff
@@ -341,6 +347,20 @@ const generateFilteredTestcases = (): Testcase[] => {
     environment: env,
   });
 
+  // Stamp the run-scoped trajectory group so every task in this run lands under
+  // one folder (`<root>/<experiment>__<model>__<runToken>/...`) instead of
+  // scattered per-task timestamps. The run token is generated ONCE here and
+  // reused at completion time, so re-running the same suite can't clobber the
+  // previous run's experiment.json. Local persistence only — does not affect
+  // Braintrust.
+  const trajectoryGroup = buildTrajectoryGroupSlug({
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/evals/index.eval.ts]: Partly addressed in `0b8c03e7` — with a deliberate deviation worth flagging.

The risky derivation logic is now directly unit-tested rather than via a mocked-Eval harness: `resolveUnambiguousModel` (single/two/core-only/empty/duplicate/mixed), the slug composition, run-token format + anti-collision, the persistence gate, the group-dir-existence backstop, and the metadata source. Every one is mutation-tested — each fails if its fix is reverted.

I did not add a mocked-`Eval` integration test for `index.eval.ts`: it is a single top-level IIFE that runs on import, so covering it means faking argv parsing, the task registry and the dataset loaders — high-maintenance for a legacy entrypoint being superseded by `framework/runner.ts` (which already has `core-runner.test.ts` / `braintrust-runner-nolog.test.ts`). Happy to add it if you disagree.

Note the lifecycle did change in this commit: legacy testcases are now materialized before `Eval()` (the group's model can only be derived from them, and the stamp must precede any recorder construction), kept inside the existing `try` so a dataset-load failure is still reported instead of escaping as an unhandled rejection. Braintrust invokes `data()` exactly once (`braintrust/dist/index.js:11919`), so returning the cached array is equivalent — and strictly more deterministic, since `generateFilteredTestcases` routes through random sampling.

File: packages/evals/framework/runner.ts [hunk 11]
```diff
@@ -343,6 +349,23 @@ export async function runEvals(
...
+  // is generated ONCE here and reused for the completion-time experiment link, so
+  // a re-run of the same suite can't clobber the previous run's experiment.json.
+  // Local persistence only — does not affect Braintrust experiment naming.
+  const trajectoryGroup = buildTrajectoryGroupSlug({
+    experimentName,
+    model: options.modelOverride,
+    runToken: generateRunToken(),
+  });
+  process.env.EVAL_EXPERIMENT_NAME = experimentName;
+  process.env.EVAL_TRAJECTORY_GROUP = trajectoryGroup;
```
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/evals/framework/runner.ts:365]: Partly valid — the concrete leak is fixed in `0b8c03e7`; the concurrency case is not reachable and I have left it.

The real bug your comment points at: the entrypoints stamp `EVAL_TRAJECTORY_GROUP` / `EVAL_EXPERIMENT_NAME` / `EVAL_TRAJECTORY_MODEL` straight onto `process.env`, but `withEnvOverrides` (`tui/commands/parse.ts`) restored only the keys *it* set — so a run's group survived the REPL command that created it, contradicting that function's own docstring about not leaking between commands. It now snapshots and restores the run-stamped keys too, which fixes both call sites without touching them. Tests cover the leak and the restore-to-prior-value case.

On carrying the group in run/task context instead of env: agreed in principle, but concurrent `runEvals()` is not reachable today — `tui/commands/run.ts:289` awaits a single call, and every recorder is constructed inside the Braintrust `task` callback, strictly after the stamp. Threading it through would mean changing `TrajectoryRecorder`, `verifierAdapter` and every bench task that calls `runWithVerifier`. That is a bigger refactor than this PR should carry; with the leak closed, the remaining exposure is a latent design constraint rather than a live defect.

File: packages/evals/framework/trajectoryGroup.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/evals/framework/trajectoryGroup.ts]: (Replying here since the entropy thread is on an older commit.)

Valid math — confirmed ~2.94% at n=1000 — and fixed in the latest commit by widening to 8 bytes (2^64), which drops even the 1000-simultaneous-start case to ~3e-14. The premise is unrealistic (real concurrency is 2–8 runs, i.e. ~0.0002% at 3 bytes), but the width is free, so I sized it for the pathological case rather than argue the odds.

I did **not** take the alternative of atomically reserving the group directory, because it would regress a fix from the previous round. Reserving the group up front requires the entrypoint to create `<root>/<group>/` before any task runs — which defeats `writeExperimentLink`'s "group dir exists ⇒ something was recorded" backstop and brings back the empty group dirs that were just eliminated for core-only and non-agent-category runs. Reservation stays where it can be atomic without that cost: the per-trajectory leaf (`reserveTrajectoryDir`).

Also note the pre-existing token tests injected entropy, so they could not have caught the default width shrinking — added one asserting the real default (16 hex chars) plus distinctness across 200 calls. It fails if reverted to 3 bytes.

File: packages/evals/framework/trajectoryRecorder.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (miguelg719) [reply packages/evals/framework/trajectoryRecorder.ts:291]: Re-checked this with a test rather than leaving it at my earlier reading — pushed in the latest commit.

The test runs `persistResult()` **before** `finish()` (the reverse of the adapter's order) and asserts the score survives, every trajectory file is present, and the reservation did not split. **It passes**, so the reported data loss does not reproduce:

- `persistResult()` writes `<dir>/scores/result.json`.
- `finish()` → `writeTrajectoryDir()` writes `task_data.json`, `trajectory.json`, `screenshots/` — and never touches `scores/`.
- The `result` keys `writeTrajectoryDir` does write are `step.toolOutput.result` (per-step tool output), unrelated to the `EvaluationResult` that `persistResult()` persists.

So there is no merge/re-apply needed. Keeping the test regardless of the disagreement: the PR advertises that result persistence is order-free, and an advertised ordering guarantee with no test is one refactor away from being false. Happy to be shown a concrete failing sequence if you think I have missed one.

**Review by seanmcguire12 — APPROVED** [review]
